HeterogeneousCore/CUDACore/README.md

```diff
@@ -0,0 +1,626 @@
+  void produce(edm::Event& iEvent, edm::EventSetup const& iSetup) override;
+  ...
+private:
+  ...
+  ProducerOutputGPUAlgo gpuAlgo_;
+  edm::EDGetTokenT<InputData> inputToken_;
+  edm::EDPutTokenT<CUDAProduct<OutputData>> outputToken_;
+  CUDAContextToken ctxTmp_;
+};
+...
+void ProducerOutputCUDA::acquire(edm::Event const& iEvent, edm::EventSetup const& iSetup, edm::WaitingTaskWithArenaHolder waitingTaskHolder) {
+  // Sets the current device and creates a CUDA stream
+  CUDAScopedContext ctx{iEvent.streamID(), std::move(waitingTaskHolder)};
+
+  auto const& inputData = iEvent.get(inputToken_);
+
+  // Queues asynchronous data transfers and kernels to the CUDA stream
+  // returned by CUDAScopedContext::stream()
+  gpuAlgo.makeAsync(inputData, ctx.stream());
+
+  // Passes the current device and CUDA stream to produce()
+  // Feels a bit silly, and will hopefully get improved in the future
+  ctxTmp_ = ctx.toToken();
+
+  // Destructor of ctx queues a callback to the CUDA stream notifying
+  // waitingTaskHolder when the queued asynchronous work has finished
+}
+
+// Called after the asynchronous work has finished
+void ProducerOutputCUDA::produce(edm::Event& iEvent, edm::EventSetup const& iSetup) {
+  // Sets again the current device, uses the CUDA stream created in the acquire()
+  CUDAScopedContext ctx{std::move(ctxTmp_)};
+
+  // Now getResult() returns data in GPU memory that is passed to the
+  // constructor of OutputData. CUDAScopedContext::emplace() wraps the
+  // OutputData to CUDAProduct<OutputData>. CUDAProduct<T> stores also
+  // the current device and the CUDA stream since those will be needed
+  // in the consumer side.
+  ctx.emplace(iEvent, outputToken_, gpuAlgo.getResult());
+}
+```
+
+### Producer with CUDA input
+
+```cpp
+class ProducerInputCUDA: public edm::stream::EDProducer<ExternalWork> {
+public:
+  ...
+  void acquire(edm::Event const& iEvent, edm::EventSetup const& iSetup, edm::WaitingTaskWithArenaHolder waitingTaskHolder) override;
+  void produce(edm::Event& iEvent, edm::EventSetup const& iSetup) override;
+  ...
+private:
+  ...
+  ProducerInputGPUAlgo gpuAlgo_;
+  edm::EDGetTokenT<CUDAProduct<InputData>> inputToken_;
+  edm::EDPutTokenT<OutputData> outputToken_;
+};
+...
+void ProducerInputCUDA::acquire(edm::Event const& iEvent, edm::EventSetup& iSetup, edm::WaitingTaskWithArenaHolder waitingTaskHolder) {
+  CUDAProduct<InputData> const& inputDataWrapped = iEvent.get(inputToken_);
+
+  // Set the current device to the same that was used to produce
+  // InputData, and also use the same CUDA stream
+  CUDAScopedContext ctx{inputDataWrapped, std::move(waitingTaskHolder)};
+
+  // Alternatively, if e.g. there is another module queuing
+  // independent work to the CUDA stream, a new CUDA stream can also be
+  // created here with
+  CUDAScopedContext ctx{iEvent.streamID(), std::move(waitingTaskHolder);
```

Comment:
Could you elaborate a bit more on this use case ?

Also, how do we handle the case of reading two CUDA objects, produced by independent modules ?

Reply:
> Could you elaborate a bit more on this use case ?

I can try. Essentially this is for a case where there are two (or more) consumers of `CUDAProduct<InputData>` whose work is independent and thus can be run in parallel.

Ok, writing this down makes me think that deciding here whether or not to create a CUDA stream is a rather low-level concept, and that maybe the CUDA streams should be managed by the framework somehow when the full DAG of modules is known. But that would be way beyond this PR. Or I(/we) give up on the "streaming_node" mode.

> Also, how do we handle the case of reading two CUDA objects, produced by independent modules ?

I'll add an example (essentially just
```cpp
  auto const& otherInputData = ctx.get(iEvent.get(otherInputToken_));
  // or
  auto const& otherInputData = ctx.get(iEvent, otherInputToken_);
```
)

Reply:
> I can try. Essentially this is for a case where there are two (or more) consumers of `CUDAProduct<InputData>` whose work is independent and thus can be run in parallel.

Ah,  I think I understand: you mean to avoid that multiple `EDProducer`s queue work in the same CUDA stream in parallel ?
If so, I understand this is something we may need - but I am not sure if it is something that the individual `EDProducer` can know a priori, right ?

Reply:
> Ah, I think I understand: you mean to avoid that multiple `EDProducer`s queue work in the same CUDA stream in parallel ?

Yes.

> If so, I understand this is something we may need - but I am not sure if it is something that the individual `EDProducer` can know a priori, right ?

That's exactly how I feel to think as well. But I'd say let's start with this, and think about improvements. (I'll anyway clarify the README)

Reply:
> > If so, I understand this is something we may need - but I am not sure if it is something that the individual `EDProducer` can know a priori, right ?
> 
> That's exactly how I feel to think as well. But I'd say let's start with this, and think about improvements. (I'll anyway clarify the README)

Created an issue #278.